core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -834,7 +837,9 @@ private[spark] class BlockManager(
 
     // Either we're storing bytes and we asynchronously started replication, or we're storing
     // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
+    // Should not replicate the block if putBlockStatus is StorageLevel.NONE
+    if (putBlockStatus != null &&
+      putBlockStatus.storageLevel != StorageLevel.NONE && putLevel.replication > 1) {
```

Comment:
+1
I like limiting the side effects to `putBlockInfo`.  (though tiny nit, I think you forgot a `!`, should be

 `if (!putBlockInfo.isFailed && ...`